### PR TITLE
[feature](nereids) Plan Translator

### LINKED ISSUE #9621: [Enhancement](Nereids)Add physical plan nodes that needed by SSB
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

In SSB, we need these type nodes:
- Filter
- Aggregate
- Join(Only need inner join, but we should add all join type in these stage)
- Sort(Order By)
- Project

### Solution

We need to check out what attributes we need to add in each node according to PlanNode in Fragment.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

Issue Number: close #9621

## Problem Summary:

Add following physical operator: PhysicalAgg PhysicalSort PhysicalHashJoin

Add basic logic of plan translator

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No)
3. Has document been added or modified: (No)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PhysicalPlanTranslator.java [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+                aggregationNode = new AggregationNode(context.nextNodeId(), inputPlanFragment.getPlanRoot(), aggInfo);
+                aggregationNode.unsetNeedsFinalize();
+                aggregationNode.setUseStreamingPreagg(physicalAggregation.isUsingStream());
+                aggregationNode.setIntermediateTuple();
+                if (!partitionExpressionList.isEmpty()) {
+                    inputPlanFragment.setOutputPartition(DataPartition.hashPartitioned(execPartitionExpressions));
+                }
+                break;
+            case FIRST_MERGE:
+                aggInfo = AggregateInfo.create(execGroupingExpressions, execAggExpressions, outputTupleDesc,
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PhysicalPlanTranslator.java:116]: do we need exchange node under agg merge node?

File: fe/fe-core/src/main/java/org/apache/doris/planner/ExchangeNode.java [hunk 2]
```diff
@@ -64,6 +64,8 @@ public class ExchangeNode extends PlanNode {
     // only if mergeInfo_ is non-null, i.e. this is a merging exchange node.
     private long offset;
 
+    private DataPartition dataPartition;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/planner/ExchangeNode.java:67]: why add this?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+public class PhysicalSort extends PhysicalUnaryOperator<PhysicalSort, PhysicalPlan> {
+
+    private int offset;
+
+    private int limit = -1;
+
+    private List<OrderKey> orderList;
+
+    // if true, the output of this node feeds an AnalyticNode
+    private boolean isAnalyticSort;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java:35]: current, we do not support analytic functions. In old planner, this is a trick flag. In Nereids, this should no longer exists

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalOperator.java [hunk 4]
```diff
@@ -29,4 +29,6 @@
  */
 public interface PhysicalOperator<TYPE extends PhysicalOperator<TYPE>> extends PlanOperator<TYPE> {
     List<Slot> computeOutputs(LogicalProperties logicalProperties, Plan... inputs);
+
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalOperator.java:32]: blank lines

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalOlapScan.java [hunk 5]
```diff
@@ -34,6 +35,8 @@ public class PhysicalOlapScan extends PhysicalScan<PhysicalOlapScan> {
     private final List<Long> selectedTabletId;
     private final List<Long> selectedPartitionId;
 
+    private OlapTable olapTable;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalOlapScan.java:38]: why move this from parent class to sub class

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+public class PhysicalAggregation extends PhysicalUnaryOperator<PhysicalAggregation, PhysicalPlan>{
+
+    private List<Expression> groupByExprList;
+
+    private List<Expression> aggExprList;
+
+    private List<Expression> partitionExprList;
+
+    private AggregateInfo.AggPhase aggPhase;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java:18]: maybe a new enum is better

**Comment by morrySnow** [comment]
fix check style~

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PhysicalPlanTranslator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PhysicalPlanTranslator.java:116]: exactly

File: fe/fe-core/src/main/java/org/apache/doris/planner/ExchangeNode.java (hunk 2 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/ExchangeNode.java:67]: removed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/Plan.java [hunk 7]
```diff
@@ -44,4 +44,6 @@
 
     @Override
     Plan child(int index);
+
```
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/Plan.java:47]: removed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java (hunk 3 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java:35]: removed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalOperator.java (hunk 4 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalOperator.java:31]: removed

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java (hunk 6 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java:18]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PlanContext.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    private final IdGenerator<PlanFragmentId> fragmentIdGenerator = PlanFragmentId.createGenerator();
+
+    private final IdGenerator<PlanNodeId> nodeIdGenerator = PlanNodeId.createGenerator();
+
+    public List<PlanFragment> getPlanFragmentList() {
+        return planFragmentList;
+    }
+
+    public TupleDescriptor generateTupleDesc() {
+        return analyzer.getDescTbl().createTupleDescriptor();
```
[the remark this answers is not in the document]
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PlanContext.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/planner/HashJoinNode.java [hunk 9]
```diff
@@ -140,6 +141,60 @@ public HashJoinNode(PlanNodeId id, PlanNode outer, PlanNode inner, TableRef inne
         }
     }
 
+    public HashJoinNode(PlanNodeId id, PlanNode outer, PlanNode inner, JoinOperator joinOp, List<Expr> eqJoinConjuncts,
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/planner/HashJoinNode.java:144]: Why this PR include `planner`?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java [hunk 10]
```diff
@@ -17,9 +17,20 @@
 
 package org.apache.doris.nereids.properties;
 
+import org.apache.doris.analysis.DistributionDesc;
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java:20]: config `style` according to [doc](https://doris.apache.org/zh-CN/developer-guide/java-format-code.html#import-order)

File: fe/fe-core/src/main/java/org/apache/doris/planner/HashJoinNode.java (hunk 9 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/planner/HashJoinNode.java:144]: > Why this PR include `planner`?

And any other modification to Planner is based on the similiar reason as commented above

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java (hunk 10 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/PhysicalProperties.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/AbstractOperator.java [hunk 11]
```diff
@@ -17,13 +17,18 @@
...
+import org.apache.doris.nereids.trees.plans.physical.PhysicalPlan;
+
 import java.util.Objects;
 
 /**
  * Abstract class for all concrete operator.
  */
 public abstract class AbstractOperator<TYPE extends AbstractOperator<TYPE>> implements Operator<TYPE> {
     protected final OperatorType type;
+    protected long limited;
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/AbstractOperator.java:31]:
```suggestion
    protected final long limited;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+package org.apache.doris.nereids;
+
+import org.apache.doris.nereids.operators.plans.physical.PhysicalOperator;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalPlan;
+
+@SuppressWarnings("rawtypes")
+public abstract class PlanOperatorVisitor<R, C> {
+    public abstract R visit(PhysicalPlan<? extends PhysicalPlan, ? extends PhysicalOperator> physicalPlan, C context);
+
+    public R visitPhysicalAggregationPlan(PhysicalPlan<? extends PhysicalPlan, ? extends PhysicalOperator> physicalPlan,
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java:27]:
```suggestion
    public R visitPhysicalAggregation(PhysicalPlan<? extends PhysicalPlan, PhysicalAggregation> aggregationPlan, C context);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/AbstractOperator.java [hunk 13]
```diff
@@ -33,4 +38,17 @@ public AbstractOperator(OperatorType type) {
...
+
+    public <R, C> R accept(PlanOperatorVisitor<R, C> visitor, PhysicalPlan<?, ?> physicalPlan, C context) {
+        return null;
+    }
+
+    public long getLimited() {
+        return limited;
+    }
+
+    public void setLimited(long limited) {
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/AbstractOperator.java:50]: Operator and Plan's properties should be immutable. So no setter.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+import org.apache.doris.nereids.operators.OperatorType;
+import org.apache.doris.nereids.operators.plans.AggPhase;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalPlan;
+
+import java.util.List;
+
+public class PhysicalAggregation extends PhysicalUnaryOperator<PhysicalAggregation, PhysicalPlan> {
+
+    private List<Expression> groupByExprList;
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java:29]: operator's properties should be final

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalHashJoin.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.apache.doris.nereids.operators.OperatorType;
+import org.apache.doris.nereids.operators.plans.JoinType;
+import org.apache.doris.nereids.trees.expressions.ComparisonPredicate;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalPlan;
+
+public class PhysicalHashJoin extends PhysicalBinaryOperator<PhysicalHashJoin, PhysicalPlan, PhysicalPlan> {
+
+    private JoinType joinType;
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalHashJoin.java:28]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+import org.apache.doris.nereids.operators.OperatorType;
+import org.apache.doris.nereids.properties.OrderKey;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalPlan;
+
+import java.util.List;
+
+public class PhysicalSort extends PhysicalUnaryOperator<PhysicalSort, PhysicalPlan> {
+
+    private int offset;
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java:28]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+// under the License.
+
+package org.apache.doris.nereids;
+
+import org.apache.doris.nereids.operators.plans.physical.PhysicalOperator;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalPlan;
+
+@SuppressWarnings("rawtypes")
+public abstract class PlanOperatorVisitor<R, C> {
+    public abstract R visit(PhysicalPlan<? extends PhysicalPlan, ? extends PhysicalOperator> physicalPlan, C context);
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java:25]: PhysicalPlan should be Plan.
```
public abstract R visit(Plan<? extends Plan, ? extends PlanOperator> plan, C context);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids;
+
+import org.apache.doris.nereids.operators.plans.physical.PhysicalOperator;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalPlan;
+
+@SuppressWarnings("rawtypes")
+public abstract class PlanOperatorVisitor<R, C> {
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java:24]: Is this class a generic visitor for all the operators, i.e., both logical and physical, or it's a dedicated visitor only for physical operators?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java (hunk 17 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java (hunk 18 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java:26]: It's a generic visitor

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java (hunk 16 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalSort.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalHashJoin.java (hunk 15 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalHashJoin.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java (hunk 14 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/plans/physical/PhysicalAggregation.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/AbstractOperator.java (hunk 13 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/AbstractOperator.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java (hunk 12 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/PlanOperatorVisitor.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/AbstractOperator.java (hunk 11 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/AbstractOperator.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java [hunk 19]
```diff
@@ -38,49 +38,48 @@
...
  * computations.
  * TODO: get rid of this by transforming
  *   SELECT COUNT(DISTINCT a, b, ..) GROUP BY x, y, ...
  * into an equivalent query with a inline view:
  *   SELECT COUNT(*) FROM (SELECT DISTINCT a, b, ..., x, y, ...) GROUP BY x, y, ...
- *
+ * <p>
  * The tree structure looks as follows:
  * - for non-distinct aggregation:
- *   - aggInfo: contains the original aggregation functions and grouping exprs
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:52]: We'd better not change the original comment indents.
You could wrap the comments with the `<pre></pre>` HTML tag, thus the indents wouldn't be modified when formatting the code style.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java [hunk 20]
```diff
@@ -38,49 +38,48 @@
...
  */
 public final class AggregateInfo extends AggregateInfoBase {
     private final static Logger LOG = LogManager.getLogger(AggregateInfo.class);
 
     public enum AggPhase {
-        FIRST,
-        FIRST_MERGE,
-        SECOND,
-        SECOND_MERGE;
+        FIRST, FIRST_MERGE, SECOND, SECOND_MERGE;
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:75]: using the original code style?

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java [hunk 21]
```diff
@@ -277,29 +271,27 @@ public static boolean estimateIfContainsMultiDistinct(List<FunctionCallExpr> dis
...
      * At the moment, we require that all distinct aggregate
      * functions be applied to the same set of exprs (ie, we can't do something
      * like SELECT COUNT(DISTINCT id), COUNT(DISTINCT address)).
      * Aggregation happens in two successive phases:
      * - the first phase aggregates by all grouping exprs plus all parameter exprs
-     *   of DISTINCT aggregate functions
-     *
+     * of DISTINCT aggregate functions
+     * <p>
      * Example:
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:282]: using the original indents?

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java [hunk 22]
```diff
@@ -414,20 +405,20 @@ public void getRefdSlots(List<SlotId> ids) {
      * Substitute all the expressions (grouping expr, aggregate expr) and update our
      * substitution map according to the given substitution map:
      * - smap typically maps from tuple t1 to tuple t2 (example: the smap of an
-     *   inline view maps the virtual table ref t1 into a base table ref t2)
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:417]: ditto.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java [hunk 23]
```diff
@@ -539,15 +526,14 @@ private Expr createCountDistinctAggExprParam(int firstIdx, int lastIdx,
      * - 'this' is the phase 1 aggregation
      * - grouping exprs are those of the original query (param origGroupingExprs)
      * - aggregate exprs for the DISTINCT agg fns: these are aggregating the grouping
-     *   slots that were added to the original grouping slots in phase 1;
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:542]: ditto.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java [hunk 24]
```diff
@@ -733,8 +707,8 @@ public void createSmaps(Analyzer analyzer) {
      * - The parameters of the sum function may involve the columns of a materialized view.
      * - The type of this column may happen to be inconsistent with the column type of the base table.
      * - In order to ensure the correctness of the result,
-     *   the parameter type needs to be changed to the type of the materialized view column
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:736]: ditto.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java [hunk 25]
```diff
@@ -761,10 +735,10 @@ public void updateTypeOfAggregateExprs() {
      * Mark slots required for this aggregation as materialized:
      * - all grouping output slots as well as grouping exprs
      * - for non-distinct aggregation: the aggregate exprs of materialized aggregate slots;
-     *   this assumes that the output slots corresponding to aggregate exprs have already
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:764]: ditto.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java (hunk 19 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java:52]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java (hunk 21 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java (hunk 20 above)
Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/AggregateInfo.java]: done

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**